### ISSUE TITLE
regarding the backpropagation with nms

### ISSUE THREAD

**Issue by tao-bai** [ianmcdiarmidsterling/adversarial#2]
I thought nms would break the computation graph and lead to the failure of back  propagation. But in your code, it works. I couldn't figure it out. Could you please explain a little bit?

**Comment by IMSterling** [comment]
What exactly do you mean by "break the
computation graph"?

In my code, the backpropagation is training an adversarial patch, using a
fixed neural network. In this case the neural network (YoloV3) does indeed
use nms during a forward pass. However for our purposes this does not
matter and we can treat the fundamental neural network as a black box,
because it is fixed. All we need to know is that we can compute a gradient
when we pass an image into the trained network and receive a confidence
score out. From this gradient, we can then use back propagation on the
adversarial patch to take an optimization step.

Because we are not modifying the network, we do not need to worry about nms
interfering with the backpropagation.

As a side note: if you are considering doing optimization, DO NOT USE MY
OPTIMIZE ROUTINE, as it performs a grid search and this is a computational
nightmare. Instead, I would consider optimizing with a package like Optuna.

Best
Ian

**Comment by tao-bai** [comment]
Hi Ian! Thanks for your detailed explanation! I understand the idea of treating the model as a black box and fix it in my codes. However, it doesn't work at this moment. Anyway, thank you for your great work so much!

**Comment by IMSterling** [comment]
Closed (time)
